cartoframes/viz/kuviz.py

```diff
@@ -22,48 +31,72 @@ def all(credentials=None):
         kuvizs = km.all()
         return [kuviz_to_dict(kuviz) for kuviz in kuvizs]
 
-    def set_credentials(self, credentials=None):
-        self._credentials = credentials or get_default_credentials()
-        self._auth_client = _create_auth_client(self._credentials)
-
-    def publish(self, html, name, password=None):
-        return _create_kuviz(html=html, name=name, auth_client=self._auth_client, password=password)
-
-    def is_sync(self):
-        return all(layer.source.dataset.is_remote() for layer in self._layers)
+    def get_layers(self):
+        return self._layers
 
-    def is_public(self):
-        return all(layer.source.dataset.is_public() for layer in self._layers)
+    def set_layers(self, layers, name, table_name):
+        self._sync_layers(layers, table_name)
+        self._manage_maps_api_key(name)
+        self._add_layers_credentials()
 
-    def get_layers(self, maps_api_key='default_public'):
+    def publish(self, html, name, password=None):
+        self.kuviz = _create_kuviz(html=html, name=name, auth_client=self._auth_client, password=password)
+        return kuviz_to_dict(self.kuviz)
+
+    def update(self, data, name, password):
+        if not self.kuviz:
+            raise CartoException('The map has not been published. Use the `publish` method.')
+
+        self.kuviz.data = data
+        self.kuviz.name = name
+        self.kuviz.password = password
+        self.kuviz.save()
+
+        return kuviz_to_dict(self.kuviz)
+
+    def delete(self):
+        if self.kuviz:
+            self.kuviz.delete()
+            warn("Publication '{n}' ({id}) deleted".format(n=self._publisher.kuviz.name, id=self._publisher.kuviz.id))
+            self.kuviz = None
+            return True
+        return False
+
+    def _sync_layers(self, layers, table_name):
+        for idx, layer in enumerate(layers):
+            if layer.source.dataset.is_local():
+                table_name = normalize_name("{name}_{idx}".format(name=table_name, idx=idx))
+                layer = self._sync_layer(layer, table_name)
+            self._layers.append(layer)
+
+    def _sync_layer(self, layer, table_name):
+        layer.source.dataset.upload(table_name=table_name, credentials=self._credentials)
+        layer.source = Source(table_name, credentials=self._credentials)
+        warn('Table `{}` created. In order to publish the map, you will need to create a new Regular API '
```

Comment:
I think we should warn the client about that, don't you?

Reply:
If you publish a map with several layers, you'd get a bunch of warnings which actually might seem errors. 

I think we could at most show the message once with a summary. It would be much better using a logger as well (but we are not there yet).

In any case, isn't the whole purpose of this PR to create the API key automatically? Why we still need to show this message?

Reply:
> In any case, isn't the whole purpose of this PR to create the API key automatically? Why we still need to show this message?

You are right, the message is wrong.

Thinking about the rest, we are showing some messages (table creation), but not others (API key creation). Furthermore, our current messages seem like errors, so we could avoid them by now. 

I have added a comment about that in the logging issue: https://github.com/CartoDB/cartoframes/issues/422#issuecomment-541422026 and I am going to remove the current ones.